### PR TITLE
melt/payoutRewards before updating ratio

### LINKED ISSUE #595: Fix inconsistent melting behavior by melting before any parameter change

### PR DESCRIPTION
**Pull request by pmckelvy1**
fixes #595 
fixes https://github.com/code-423n4/2023-01-reserve-findings/issues/380
fixes https://github.com/code-423n4/2023-01-reserve-findings/issues/287

### CODE REVIEW

File: contracts/p0/Furnace.sol [hunk 1]
```diff
@@ -50,6 +50,7 @@ contract FurnaceP0 is ComponentP0, IFurnace {
     /// Ratio setting
     /// @custom:governance
     function setRatio(uint192 ratio_) public governance {
+        melt();
```
Comment (tbrent) [inline contracts/p0/Furnace.sol:53]: Governance may want to change the ratio while paused or frozen.  Currently the `melt()` call will revert. I think we want a try-catch here.

Comment (pmckelvy1) [reply contracts/p0/Furnace.sol:53]: good catch

File: contracts/p0/StRSR.sol [hunk 2]
```diff
@@ -524,6 +524,7 @@ contract StRSRP0 is IStRSR, ComponentP0, EIP712Upgradeable {
     }
 
     function setRewardRatio(uint192 val) public governance {
+        _payoutRewards();
```
Comment (tbrent) [inline contracts/p0/StRSR.sol:527]: Here we want to do a similar thing and only payout rewards if not frozen

Comment (tbrent) [reply contracts/p0/StRSR.sol:527]: Possibly cleanest is making `payoutRewards()` public and calling it from here.

Comment (tbrent) [reply contracts/p0/StRSR.sol:527]: I can handle this, though, once the tests pass

Comment (pmckelvy1) [reply contracts/p0/StRSR.sol:527]: the internal `_payoutRewards()` does not care if the system is paused or frozen, only the external call has the modifier

Comment (tbrent) [reply contracts/p0/StRSR.sol:527]: Right but we should not hand out any rewards while frozen, even if governance changes the ratio

Comment (tbrent) [reply contracts/p0/StRSR.sol:527]: Sorry, my previous suggestion was bad. We should just do `if (!main.frozen()) _payoutRewards()` and nothing more, I think

**Pull request merged**